include/tvm/runtime/profiling.h

```diff
@@ -0,0 +1,162 @@
+   * them before calling `SyncAndGetElapsedNanos` on any of them.
+   *
+   * Note: this function should be only called once per object. It may incur
+   * a large synchronization overhead (for example, with GPUs).
+   */
+  virtual int64_t SyncAndGetElapsedNanos() = 0;
+
+  virtual ~TimerNode() {}
+
+  static constexpr const char* _type_key = "TimerNode";
+  TVM_DECLARE_BASE_OBJECT_INFO(TimerNode, Object);
+};
+
+/*! \brief Timer for a specific device.
+ */
+class Timer : public ObjectRef {
+ public:
+  /*!
+   * \brief Get a device specific timer.
+   * \param ctx The device context to time.
+   * \return A `Timer` that has already been started.
+   *
+   * Use this function to time runtime of arbitrary regions of code on a specific
+   * device. The code that you want to time should be running on the device
+   * otherwise the timer will not return correct results. This is a lower level
+   * interface than TimeEvaluator and only runs the timed code once
+   * (TimeEvaluator runs the code multiple times).
+   *
+   * A default timer is used if a device specific one does not exist. This
+   * timer performs synchronization between the device and CPU, which can lead
+   * to overhead in the reported results.
+   *
+   * Example usage:
+   * \code{.cpp}
+   * Timer t = Timer::Start(TVMContext::cpu());
+   * my_long_running_function();
+   * t.Stop();
+   * ... // some more computation
+   * int64_t nanosecs = t.SyncAndGetElapsedNanos() // elapsed time in nanoseconds
+   * \endcode
+   *
+   * To add a new device-specific timer, register a new function
+   * "profiler.timer.my_device" (where `my_device` is the `DeviceName` of your
+   * device). This function should accept a `TVMContext` and return a new `Timer`
+   * that has already been started.
+   *
+   * For example, this is how the CPU timer is implemented:
+   * \code{.cpp}
+   *  class CPUTimerNode : public TimerNode {
+   *   public:
+   *    virtual void Start() { start_ = std::chrono::high_resolution_clock::now(); }
+   *    virtual void Stop() { duration_ = std::chrono::high_resolution_clock::now() - start_; }
+   *    virtual int64_t SyncAndGetElapsedNanos() { return duration_.count(); }
+   *    virtual ~CPUTimerNode() {}
+   *
+   *    static constexpr const char* _type_key = "CPUTimerNode";
+   *    TVM_DECLARE_FINAL_OBJECT_INFO(CPUTimerNode, TimerNode);
+   *
+   *   private:
+   *    std::chrono::high_resolution_clock::time_point start_;
+   *    std::chrono::duration<int64_t, std::nano> duration_;
+   *  };
+   *  TVM_REGISTER_OBJECT_TYPE(CPUTimerNode);
+   *
+   *  TVM_REGISTER_GLOBAL("profiling.timer.cpu").set_body_typed([](TVMContext ctx) {
+   *    return Timer(make_object<CPUTimerNode>());
+   *  });
+   * \endcode
+   */
+  static TVM_DLL Timer Start(TVMContext ctx);
+
+  /*! \brief Stop the timer.
+   *
+   * Note: this function should only be called once per object.
+   */
+  void Stop() { operator->()->Stop(); }
```

Comment:
consider remove these member functions. 
Given we already have timer->Stop() and timer->SyncAndGetElapsedNanos() to avoid one level of indirection.

This is also the approaches we use in latest added objects

Reply:
I like including these methods as they document the timer interface.

Reply:
these methods are also presented in the node and documentations are there as well. Removing the methods indirection would make the code more consistent with the reference usage of the rest of the codebase.

It also reduces the confusion of two possible ways to do the same thing.

Adding a link \sa link, as well as code examples in the Timer class would resolved the usage problem.

Reply:
I based all of this off of how `Pass` does things. And it takes this same approach.

The API here is `Timer` not `TimerNode`. These methods are the interface. Having these convenience methods also provides autocompletion for people who rely on it.